### PR TITLE
Switched Sockets perf tests to Benchmarks

### PR DESCRIPTION
**Pull request by ianhays**
Rough draft of what the sockets tests will look switched to use Benchmarks for the perf tests. Ideally there would only be one tests csproj as in this PR and tests would function as they do in other projects.

These perf tests run on Windows and give reasonable results, but on Linux they throw an error that "ws2_32.dll" is missing.

@pgavlin @stephentoub

### CODE REVIEW

**Comment by ianhays** [comment]
What is the purpose of the System.Net.Sockets.Legacy tests? Should those perf tests be updated as these will be?

**Comment by davidsh** [comment]
cc: @CIPop

File: src/Common/tests/System.Net/Sockets/Performance/SocketPerformanceTests.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@
 using Xunit;
 using Xunit.Abstractions;
 
-namespace System.Net.Sockets.Performance.Tests
+namespace System.Net.Sockets.Tests
```
Comment (CIPop) [inline src/Common/tests/System.Net/Sockets/Performance/SocketPerformanceTests.cs]: Our naming convention is very clear: Unit.Tests, Functional.Tests and Performance.Tests. 
Renaming them would make their purpose unclear. Please revert these changes.

File: src/System.Net.Sockets/System.Net.Sockets.sln [hunk 2]
```diff
@@ -13,13 +13,7 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.Net.Sockets", "ref\S
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.Net.Sockets", "src\System.Net.Sockets.csproj", "{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}"
 EndProject
-Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "FunctionalTests", "FunctionalTests", "{D4227E7F-1EB9-4804-803F-D0C76B0C2BAC}"
```
Comment (CIPop) [inline src/System.Net.Sockets/System.Net.Sockets.sln]: I prefer having the projects separate since they have separate purposes.

File: src/System.Net.Sockets/System.Net.Sockets.sln [hunk 3]
```diff
@@ -57,10 +51,10 @@ Global
 		{9E212427-18B3-4EF4-966C-ED18AAC08422}.Windows_Release|Any CPU.Build.0 = Debug|Any CPU
 		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.Debug|Any CPU.ActiveCfg = Windows_Debug|Any CPU
 		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.Debug|Any CPU.Build.0 = Windows_Debug|Any CPU
-		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.FreeBSD_Debug|Any CPU.ActiveCfg = FreeBSD_Debug|Any CPU
-		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.FreeBSD_Debug|Any CPU.Build.0 = FreeBSD_Debug|Any CPU
-		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.FreeBSD_Release|Any CPU.ActiveCfg = FreeBSD_Release|Any CPU
-		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.FreeBSD_Release|Any CPU.Build.0 = FreeBSD_Release|Any CPU
+		{43311AFB-D7C4-4E5A-B1DE-855407F90D1B}.FreeBSD_Debug|Any CPU.ActiveCfg = Windows_Release|Any CPU
```
Comment (CIPop) [inline src/System.Net.Sockets/System.Net.Sockets.sln]: Probably some mistake?

File: src/System.Net.Sockets/System.Net.Sockets.sln [hunk 4]
```diff
@@ -95,26 +89,6 @@ Global
...
 		{8CBA022C-635F-4C8D-9D29-CD8AAC68C8E6}.Windows_Release|Any CPU.Build.0 = Release|Any CPU
-		{BB5C85AD-C51A-4903-80E9-6F6E1AC1AD34}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{BB5C85AD-C51A-4903-80E9-6F6E1AC1AD34}.Debug|Any CPU.Build.0 = Debug|Any CPU
-		{BB5C85AD-C51A-4903-80E9-6F6E1AC1AD34}.FreeBSD_Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{BB5C85AD-C51A-4903-80E9-6F6E1AC1AD34}.FreeBSD_Debug|Any CPU.Build.0 = Debug|Any CPU
-		{BB5C85AD-C51A-4903-80E9-6F6E1AC1AD34}.FreeBSD_Release|Any CPU.ActiveCfg = Release|Any CPU
-		{BB5C85AD-C51A-4903-80E9-6F6E1AC1AD34}.FreeBSD_Release|Any CPU.Build.0 = Release|Any CPU
-		{BB5C85AD-C51A-4903-80E9-6F6E1AC1AD34}.Linux_Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{BB5C85AD-C51A-4903-80E9-6F6E1AC1AD34}.Linux_Debug|Any CPU.Build.0 = Debug|Any CPU
-		{BB5C85AD-C51A-4903-80E9-6F6E1AC1AD34}.Linux_Release|Any CPU.ActiveCfg = Release|Any CPU
```
Comment (CIPop) [inline src/System.Net.Sockets/System.Net.Sockets.sln]: Why were these removed?

**Comment by CIPop** [comment]
There is ongoing work on this code that involves contract changes and implementation changes. Marking this as blocked for now: please do not merge.
/cc: @josguil @pgavlin

File: src/Common/tests/System.Net/Sockets/Performance/SocketPerformanceTests.cs (hunk 1 above)
Comment (davidsh) [inline src/Common/tests/System.Net/Sockets/Performance/SocketPerformanceTests.cs]: FWIW, There was a discussion about the NAMESPACES to test classes and methods.  It was agreed that ALL namespaces for test classes and methods would be <namespace of classes/methods under test>.Tests

So, for example, the HttpWebRequest API which is under the System.Net namespace but yet packaged in the System.Net.Requests namespace would have the test classes and methods under the "System.Net.Tests" namespace.

See agreement in #2898.

And System.Net.Requests changes in #3660.

Note:  The test binary name still can be named System.Net.Sockets.Unit.Tests.dll or System.Net.Sockets.Performance.Tests.dll.

cc: @stephentoub

Comment (CIPop) [inline src/Common/tests/System.Net/Sockets/Performance/SocketPerformanceTests.cs]: That doesn't really work for us as we want clear separation between the test types.

**Comment by ianhays** [comment]
No worries, this is just a rough draft to demonstrate the new xunit-perf style and get the ball rolling. The goal is to make use of existing tests as well as create new tests that conform to the new standard such that we can automate performance testing in the CI in the near future (~November). 

I separated these tests from my main test update PR (#3764) with the expectation that ya'll would probably want to implement the Net tests yourselves. If I was correct and that is the case, feel free to close this PR and please keep me looped in on your progress. If not, I'd like to sync up to ask a few questions before updating and continuing.

**Comment by stephentoub** [comment]
@ianhays, what's the plan with this PR?  Should we close it out until a plan is established, or is there already one?

**Comment by CIPop** [comment]
> @ianhays, what's the plan with this PR? Should we close it out until a plan is established, or is there already one?

To clarify: System.Net has no understanding of what this is and no plan to change the current test organization. I'm against unifying all tests into a single csproj: it makes it much harder to navigate, understand and port to other platforms such as our Desktop functional test suite.

**Comment by ianhays** [comment]
We’re working on adding new performance tests that all conform to the same standard that is built on top of xunit. There are existing perf tests (such as those for System.Net) that ideally should conform to this standard so that they can be run alongside the new tests. The plan is to have dedicated perf machines that track the performance impact over time as well as the performance impact of individual commits/PRs such that we can potentially block commits that contain perf regressions. The perf infra is based off of https://github.com/Microsoft/xunit-performance/ and some introductory instructions on test structure are available at https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/performance-tests.md. 

The structure of the test organization is not the primary purpose of the PR; it is the changes to the tests themselves that are my primary concern and what I am really looking to get feedback for. Per the suggestion of @pgavlin I am going to close this PR and resubmit it without <b>any</b> changes to the organization of the existing tests.
